### ISSUE TITLE
MX reverse mode produces NaN gradients when the analysis confidence sigmoid overflows

### ISSUE THREAD

**Issue by cdhainaut** [peterdsharpe/NeuralFoil#10]
Hi Peter,

I am using NeuralFoil inside a CasADi MX optimization model. Forward CL, CD and CM remain finite, but some reverse mode gradient entries become NaN when a vectorized case is well outside the trained alpha range.

The affected entries match confidence logits for which `exp(-x)` overflows. In the example below, the last three logits are approximately `-1.15e4`, `-1.07e6` and `-2.76e6`. The corresponding MX reverse derivatives are NaN. MX forward mode and a graph built directly with SX remain finite and agree to about `1e-16`.

Changing only

```python
1 / (1 + exp(-x))
```

to the equivalent form

```python
0.5 * (1 + tanh(x / 2))
```

makes the MX reverse derivatives finite. Replacing the sigmoid by either the identity or zero also removes the NaNs, which isolates this operation as the cause. A likely numerical mechanism is a `0 * inf` product while a zero adjoint is propagated through the overflowing exponential. I have not reduced CasADi’s internal reverse sweep further than this.

The large angles are intentional. I am not expecting useful aerodynamic predictions at `-153°`. The confidence correctly tends to zero there, and the AeroSandbox 360° wrapper evaluates the raw model before blending. The problem is simply that finite forward values can produce NaN derivatives in the wider optimization model.

Here is a standalone reproducer:

```python
import casadi as ca
import numpy as np
import aerosandbox.numpy as anp
from neuralfoil import get_aero_from_kulfan_parameters
import neuralfoil.gen2_5_architecture.main as nfm

n = 8
alpha_mx = ca.MX.sym("alpha_deg", n)
alpha_value = np.array([0., 4., 10., 15., -10., -60., -120., -153.564])
kulfan = {
    "lower_weights": [0.] * 8,
    "upper_weights": [.1, .15, .15, .12, .08, .05, .02, 0.],
    "leading_edge_weight": 0.,
    "TE_thickness": 0.,
}

def evaluate(alpha):
    return get_aero_from_kulfan_parameters(
        kulfan_parameters=kulfan,
        alpha=alpha,
        Re=1.4e6,
        model_size="medium",
    )["CD"]

cd_mx = evaluate(alpha_mx)
value = np.asarray(ca.Function("value", [alpha_mx], [cd_mx])(alpha_value))
reverse = np.asarray(ca.Function(
    "reverse", [alpha_mx], [ca.jacobian(ca.sum1(cd_mx), alpha_mx)]
)(alpha_value))
forward = np.hstack([
    np.asarray(ca.Function(
        f"forward_{j}", [alpha_mx],
        [ca.jtimes(cd_mx, alpha_mx, ca.DM(np.eye(n)[:, j:j + 1]))],
    )(alpha_value))
    for j in range(n)
])

alpha_sx = ca.SX.sym("alpha_sx", n)
cd_sx = evaluate(alpha_sx)
sx_reverse = np.asarray(ca.Function(
    "sx_reverse", [alpha_sx], [ca.jacobian(ca.sum1(cd_sx), alpha_sx)]
)(alpha_value))

original_sigmoid = nfm._sigmoid
nfm._sigmoid = lambda x: 0.5 * (1 + anp.tanh(x / 2))
alpha_fixed = ca.MX.sym("alpha_fixed", n)
cd_fixed = evaluate(alpha_fixed)
fixed_reverse = np.asarray(ca.Function(
    "fixed_reverse", [alpha_fixed],
    [ca.jacobian(ca.sum1(cd_fixed), alpha_fixed)],
)(alpha_value))
nfm._sigmoid = original_sigmoid

print("forward values finite:", np.all(np.isfinite(value)))
print("MX reverse NaN indices:", np.where(~np.isfinite(reverse.ravel()))[0])
print("MX forward finite:", np.all(np.isfinite(forward)))
print("SX reverse finite:", np.all(np.isfinite(sx_reverse)))
print("stable sigmoid MX reverse finite:", np.all(np.isfinite(fixed_reverse)))
```

It produces:

```text
forward values finite: True
MX reverse NaN indices: [5 6 7]
MX forward finite: True
SX reverse finite: True
stable sigmoid MX reverse finite: True
```

Would you be open to replacing `_sigmoid` in `gen2_5_architecture/main.py` with this form?

```python
def _sigmoid(x):
    return 0.5 * (1 + np.tanh(x / 2))
```

The same helper is present in `gen2_architecture/main.py`.

I compared both forms over 721 alphas from `-180°` to `180°`, using `model_size="medium"` and Reynolds `1.4e6`. The maximum differences were zero for CL, CD and CM, and `5.6e-17` for `analysis_confidence`. The complete AeroSandbox 360° wrapper then gives finite MX reverse derivatives and agrees with central finite differences to `3.6e-7` relative.

I can open a PR with the change and a regression test if useful.

**Comment by peterdsharpe** [comment]
This makes sense! Yes, if you could open that PR with change + regression test, that'd be great - thanks!
